metaflow/plugins/kfp/kfp.py

```diff
@@ -97,7 +97,32 @@ def _command(self, code_package_url, environment, step_name, step_cli):
         commands.extend(environment.bootstrap_commands(step_name))
         commands.append("echo 'Task is starting.'")
         commands.extend(step_cli)
-        return " && ".join(commands)
+        subshell_commands = " && ".join(
+            commands
+        )  # run inside subshell to capture all stdout/stderr
+        # redirect stdout/stderr to separate files, using tee to display to UI
+        redirection_commands = "> >(tee -a 0.stdout.log) 2> >(tee -a 0.stderr.log >&2)"
+
+        # Creating a template to save logs to S3. This is within a function because
+        # datastore_root is not available within the scope of this function, and needs
+        # to be provided in the `step_op` function. f strings (AFAK) don't support
+        # insertion of only a partial number of placeholder strings.
+        def create_log_cmd(log_file):
+            save_logs_cmd_template = (
+                f"python -m awscli s3 cp {log_file} {{datastore_root}}/"
+                f"{self.flow.name}/{{run_id}}/{step_name}/"
+                f"{task_id}/{log_file}"
+            )
+            return save_logs_cmd_template
+
+        log_stdout_cmd = create_log_cmd(log_file="0.stdout.log")
+        log_stderr_cmd = create_log_cmd(log_file="0.stderr.log")
+        save_logs_cmd = f"{log_stderr_cmd} >/dev/null && {log_stdout_cmd} >/dev/null"
+
+        # After the subshell/redirection commands, we capture the exit code because otherwise,
+        # due to the ';', the Popen process will always return an exit code of 0. After capturing
+        # the code, # we exit with this code manually (even if no errors are present).
+        return f"({subshell_commands}) {redirection_commands}; export exit_code=$?; {save_logs_cmd}; exit $exit_code"
```

Comment:
Just to interject one final time to suggest that is this the right way to implement this? Implementing it using an internal container command will result in failure to report logs in the case that container itself gets killed or the shell that is executing the commands here exits for some reason.

The way Splunk (and other log brokers in Kubernetes) handle this problem is by using a [`DaemonSet`](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/) that runs on the nodes of the cluster itself and inspects the running containers and their logs. It essentially is a separate process that is watching instead of being coupled. Something like [this](https://github.com/upmc-enterprises/kubernetes-fluentd) would work out of the box. It'd likely need some investigation for how best to hook it up to the Metaflow/KFP metadata as likely the naming conventions are going to be container/pod ID based instead of experiment/run based but this approach would allow you to capture logs for rescheduled pods for example (ie. a node dies, the pod gets interrupted by a higher priority job, etc.) and from my experience during those events having logs to debug is a godsend!

Reply:
Is George setting up logs this way and how far has him gone into this? I am also leaning towards log brokers like Splunk for long term. I'd accept what's in this branch as a short term solution.

Reply:
As far as I know Akshay/George got it working last week for the core workloads (ie. the core Kubeflow operator pods). Obviously Splunk is pretty complex and not sure if it's enabled for containers/pods created from KFP and don't think it'd contain any metadata that'd link those logs to the experiment/run without additional investigation/work so obviously that's the benefit of this system is I'm guessing with the S3 pathing structure it'd be inherently easy to make that mapping.

Reply:
This solution gets us the logs for 99% of scenarios, and is cheap/simple to implement.  
This Metaflow plugin logging is to adhere to Metaflow logging standards (support `python hello.py logs <run_id>` command).  

Thank you for sharing! Using a DaemonSet would be an improvement that we can add to the backlog.

Reply:
This way of logging is vulnerable to pod eviction. Say more than request resources are used, we wouldn't be sure about it. These errors do not usually happen for prod, but happens more often in dev when resource requirement is unclear for new features, or if any bugs happened. 

I do agree with @talebzeghmi that this works in most use cases. I believe zestimate team is still using this approach (where the log transfer code is put in EMR termination script).

Reply:
Cool happy with the decision, definitely agree it fits with the 80/20 prioritization (and your assessment it should cover 99% of cases). Resolving the thread, thanks for humouring me!

Reply:
True, 99% is probably stretching it.  The k8s solution could be a K8s sidecar, but worth looking into is Metaflow sidecars 

https://github.com/zillow/metaflow/blob/master/metaflow/plugins/__init__.py#L61